server/src/test/java/org/elasticsearch/common/settings/RotatableSecretTests.java

```diff
@@ -0,0 +1,226 @@
+        assertTrue(rotatableSecret.isSet());
+        assertEquals(secret2, rotatableSecret.getSecrets().current());
+        assertEquals(secret1, rotatableSecret.getSecrets().prior());
+        assertTrue(rotatableSecret.getSecrets().priorValidTill().isAfter(Instant.now()));
+        assertTrue(
+            rotatableSecret.getSecrets().priorValidTill().isBefore(Instant.now().plusMillis(TimeValue.timeValueDays(2).getMillis()))
+        );
+
+        // rotate with expiry
+        rotatableSecret.rotate(secret3, TimeValue.timeValueMillis(1));
+        Thread.sleep(2); // ensure secret2 is expired
+        assertTrue(rotatableSecret.matches(secret3));
+        assertFalse(rotatableSecret.matches(secret1));
+        assertFalse(rotatableSecret.matches(secret2));
+        assertFalse(rotatableSecret.matches(new SecureString(randomAlphaOfLength(10))));
+        assertTrue(rotatableSecret.isSet());
+        assertEquals(secret3, rotatableSecret.getSecrets().current());
+        assertNull(rotatableSecret.getSecrets().prior());
+        assertTrue(rotatableSecret.getSecrets().priorValidTill().isBefore(Instant.now()));
+
+        // unset current and prior
+        rotatableSecret.rotate(null, TimeValue.ZERO);
+        assertFalse(rotatableSecret.matches(secret3));
+        assertFalse(rotatableSecret.matches(secret1));
+        assertFalse(rotatableSecret.matches(secret2));
+        assertFalse(rotatableSecret.matches(new SecureString(randomAlphaOfLength(10))));
+        assertFalse(rotatableSecret.isSet());
+        assertNull(rotatableSecret.getSecrets().current());
+        assertNull(rotatableSecret.getSecrets().prior());
+        assertTrue(rotatableSecret.getSecrets().priorValidTill().isBefore(Instant.now()));
+    }
+
+    public void testConcurrentReadWhileLocked() throws Exception {
+        // initial state
+        RotatableSecret rotatableSecret = new RotatableSecret(secret1);
+        assertTrue(rotatableSecret.matches(secret1));
+        assertFalse(rotatableSecret.matches(secret2));
+        assertEquals(secret1, rotatableSecret.getSecrets().current());
+        assertNull(rotatableSecret.getSecrets().prior());
+
+        boolean expired = randomBoolean();
+        CountDownLatch latch = new CountDownLatch(1);
+        TimeValue mockGracePeriod = mock(TimeValue.class);  // use a mock to force a long rotation to exercise the concurrency
+        when(mockGracePeriod.getMillis()).then((Answer<Long>) invocation -> {
+            latch.await();
+            return expired ? 0L : Long.MAX_VALUE;
+        });
+
+        // start writer thread
+        Thread t1 = new Thread(() -> rotatableSecret.rotate(secret2, mockGracePeriod));
+        t1.start();
+        assertBusy(() -> assertEquals(Thread.State.WAITING, t1.getState())); // waiting on countdown latch, holds write lock
+        assertTrue(rotatableSecret.isWriteLocked());
+
+        // start reader threads
+        int readers = randomIntBetween(0, 16);
```

Comment:
Should we have at least 1 reader here so the loop after this is guaranteed to run?